### ISSUE TITLE
Broadcast stream when no one is listening: divergent behaviors

### ISSUE THREAD

**Issue by niteshb** [dart-lang/sdk#50266]
https://github.com/niteshb/dart-bug-reports/blob/master/2022_10_21_broadcast_when_no_ones_listening/bug_2022_10_21_broadcast_when_no_ones_listening.dart

Seemingly in case of broadcast stream, when initially there is no listener, the stream buffers. This is not what it does later on. This is not what the [documentation](https://api.dart.dev/stable/2.18.3/dart-async/Stream-class.html) says or is expected.

In the code above, some events fire before the first subscriber. First subscriber ends up getting all those events nevertheless. Second subscriber attaches immediately after the first subscriber, but only get subsequent events. Later both unsubscribe and a third subscriber comes after some time. Third subscriber doesn't get events fired during the period of no subscription (as expected and documented).

As per the docs:

> When an event is fired, the listener(s) at that time will receive the event. If a listener is added to a broadcast stream while an event is being fired, that listener will not receive the event currently being fired. If a listener is canceled, it immediately stops receiving events. Listening on a broadcast stream can be treated as listening on a new stream containing only the events that have not yet been emitted when the [listen](https://api.dart.dev/stable/2.18.3/dart-async/Stream/listen.html) call occurs.

While I find expected and documented behavior when all subscriptions have been cancelled and then a new subscriber comes, it is not behaving so for the first subscriber. Seemingly it is buffering up data for the first subscriber.

**Comment by matanlurey** [comment]
You are creating a non-broadcast stream (which buffers) and then converting (really, wrapping it) `asBroadcastStream()`, in which the underlying stream buffers.

**Comment by niteshb** [comment]
Thanks! 
Changing line 62 to
```dart
final _controller = StreamController<int>.broadcast();
```
makes the program behave as documented.

**Comment by lrhn** [comment]
It's deliberate behavior, not a bug. We can discuss whether it's the most desirable behavior. I think it is.

A stream should generally not need to buffer events. It shouldn't start generating events while there is no listener, or while paused. Buffering is a fallback used when events can't be stopped from being generated, and also cannot safely just be dropped.

For a single-subscription stream, it is generally not considered safe to drop events. If you add events to a controller before there is a listener (which you shouldn't), then it has to buffer them. The broadcast stream controller is better off, since it's allowed to just throw away those events.

That's why the behavior of a `singleSubscriptionStream.asBroadcastStream()` is to not generate any events until the first listener, which means not calling `listen` until the first listener, because that's when a stream is supposed to start generating events.

If you actually want a broadcast stream, creating one directly is always better than `asBroadcastStream`. (There really isn't much good to say about `asBroadcastStream`, it provides a way to have multiple concurrent listeners on a single-subscription stream, but it does so by pretending to be a broadcast stream, even if it really isn't. The underlying stream is not designed for events getting lost.)

All in all, I think the current behavior is reasonable.

**Comment by niteshb** [comment]
Guys, I am sorry, but please, in that case, stop having misnomers. If it is deliberate and/or desirable behavior or not is secondary. 

There are two types of streams as per Dart documentation, with well defined behaviors. Unicast streams buffer, broadcast streams don't buffer. asBroadcastStream(), by nomenclature is supposed to give a broadcast stream, which shouldn't buffer. If what comes out of asBroadcastStream() is "desirable" behavior, then name it so. Please don't call it asBroadcastStream(), may be call it "asBufferedTillTheFirstSubscriberForThatSubscriberBroadcastStream()" or name it something else and document that as a third type of stream. 

Btw, it is buffered only till the first subscriber, and only for that subscriber, and that is some amazing corner-case desire. Other desirable behaviors may be:
* buffered for every subscriber
* buffered for the whole timeline of stream
* combination of the above for both unicast and broadcast streams

And as you yourself say, if someone wants a Broadcast stream, as per documentation, there is a way, if one control that code that is. So please either let asBroadcastStream() behave as per its name or rename it. Stop living in this echo chamber of deliberate/desirable behaviors which is simply self-serving and self-justifying. As per me, misnomers cause confusion, and they are bugs. 

I hope the discussion is not on desirable/deliberate behavior. Please focus on what you have named this function and does it behave according to documented behavior as per its name.

In any case, please let the world know of these desirable behaviors **clearly** in the documentation [here](https://api.flutter.dev/flutter/dart-async/Stream/asBroadcastStream.html).

**Comment by lrhn** [comment]
For the record, the documentation for `asBroadcastStream` states:

> The returned stream will subscribe to this stream when its first subscriber is added,

While the implications might not be clear, it is actually documenting the current behavior.
The implications of not listening to the source stream until the first listener on the as-broadcast stream, if the current stream is a single-subscription stream backed by a controller, is that if you add events to the controller prior to someone listening, those events will be buffered until someone listens to the stream. Because that's how such a source stream behaves until someone listens.

We probably should explicitly warn against creating a non-broadcast controller and doing `asBroadcastStream` on the stream, because then you might as well create a broadcast stream controller directly.

**Comment by niteshb** [comment]
Yeah! Also try fixing people like him and what misunderstandings they might be...
* How to create broadcast streams? In a tutorial, one might think this may be the favored standardized way to create them: https://www.youtube.com/watch?v=nQBpOIHE4eE&t=127s
* Same video a few seconds later: Data being tossed out when no one is listening: https://www.youtube.com/watch?v=nQBpOIHE4eE&t=142s
